final-project/application.py

```diff
@@ -34,19 +34,94 @@ def after_request(response):
 # configure CS50 Library to use SQLite database
 db = SQL("sqlite:///final.db")
 
+# users table
+# CREATE TABLE users (id INTEGER PRIMARY KEY NOT NULL, username TEXT UNIQUE NOT NULL, email TEXT UNIQUE NOT NULL, hash TEXT NOT NULL, first TEXT, last TEXT);
+
 @app.route("/")
 @login_required
 def index():
 
     # keep user logged in
     session.get("user_id")
 
-    """Render map."""
-    if not os.environ.get("API_KEY"):
-        raise RuntimeError("API_KEY not set")
-    return render_template("index.html", key=os.environ.get("API_KEY"))
+    return apology("TODO")
+
+
+@app.route("/categories")
+@login_required
+def categories():
+
+    # keep user logged in
+    session.get("user_id")
+
+    return apology("TODO")
+
+@app.route("/product")
+# @login_required
+def product():
+
+    # keep user logged in
+    session.get("user_id")
+
+    # get name of product
+    product = "%" + request.args.get("product") + "%"
+
+    try:
+        #Get product from the database
+        product_info = db.execute("SELECT * FROM products WHERE product_name LIKE :product", product = product)
+        # Check whether products were found
+        product_check = product_info[0]
+
+    except RuntimeError:
+        # If problem with db.execute
+        return apology("Error: We'll fix this. Please try again shortly.")
+
+    except IndexError:
+        # If no products were found
+        return apology("No products named " + request.args.get("product"))
+
+    #Send data to product template
+    return render_template("product.html", product_name = product_info[0]["product_name"], image = product_info[0]["image"], link = product_info[0]["link"], description = product_info[0]["description"], brand = product_info[0]["brand"], price = usd(product_info[0]["price"]))
+    return apology("TODO")
 
-# @app.route("/articles", methods=["GET", "POST"])
+
+@app.route("/new")
+def new():
+    """Add new product"""
+
+    # remember user
+    session.get("user_id")
+
+    if request.method == "POST":
+
+        try:
+            # Add new product
+            # TODO: add request.form.get() to get parameters
+            new_product_row = db.execute("INSERT INTO products (id, category_id, product_name, link, description, image, brand, price) VALUES (NULL, 1, :product_name, :link, :description, :image, :brand, :price)", product_name = request.form.get("product_name"), link = request.form.get("link"), description = request.form.get("description"), image = request.form.get("image"), brand = request.form.get("brand"), price = request.form.get("price"))
```

Comment:
Please make this multi-line for readability. You should ideally always keep code lines < 80 chars for other developers to quickly read code, and almost all languages support this.

Reply:
Will do

Reply:
<img width="873" alt="screen shot 2017-12-02 at 6 53 17 am" src="https://user-images.githubusercontent.com/8930410/33515187-96a36d60-d72d-11e7-8f6b-8887819858a1.png">

For some reason no matter what I do, I cannot get this syntax error to go away, when this code on multiple lines.I have tried for over an hour and read the PEP 8 Style Guide from top to bottom. 

To keep moving, I am going to keep the "render_template()" functions on single lines. If i have time at the end I will come back to this.

Reply:
🏁 done